File: 프로그래머스/LV2/게임 맵 최단거리.py
def solution(maps):
    dx = [1,0,-1,0]
    dy = [0,1,0,-1]
    h, w = len(maps), len(maps[0])
    cnt = [[0 for _ in range(w)] for _ in range(h)]
    q = [(0,0)]
    cnt[0][0] = 1
    while q:
        y, x = q.pop(0)
        for i in range(4):
            ny, nx = y + dy[i], x + dx[i]
            if 0<=nx<w and 0<=ny<h and maps[ny][nx] == 1:
                maps[ny][nx] = 0
                cnt[ny][nx] = cnt[y][x] + 1
                q.append((ny,nx))
    if cnt[h-1][w-1] == 0:
        return -1
    else:
        return cnt[h-1][w-1]

File: 프로그래머스/LV2/게임 맵 최단거리.java
import java.util.LinkedList;
import java.util.Queue;
class Solution {
    public int solution(int[][] maps) {
        int answer = bfs(maps, new int[]{0,0});
        return answer;
    }
    
    private int bfs(int[][] maps, int[] start) {
        boolean[][] check = new boolean[maps.length][maps[0].length];
        Queue<int[]> q = new LinkedList<>();
        int[] dy = {1,0,-1,0}, dx = {0,1,0,-1};
        q.add(start);
        while(!q.isEmpty()) {
            int[] p = q.poll();
            int y = p[0], x = p[1];
            check[y][x] = true;
            if(y == maps.length-1 && x == maps[0].length-1) return maps[y][x];
            for(int i = 0; i<4; i++) {
                int ny = y+dy[i], nx = x+dx[i];
                if(0<=ny && ny<maps.length && 0<=nx && nx<maps[0].length) {
                    if(maps[ny][nx] == 1 && !check[ny][nx]) {
                        q.add(new int[]{ny, nx});
                        maps[ny][nx] = maps[y][x] + 1;
                    }
                }
            }
        }
        return -1;
    }
}
